Wrapping/Generators/Python/itk/support/base.py

```diff
@@ -80,11 +80,6 @@ def itk_load_swig_module(name: str, namespace=None):
                     namespace[k] = v
         return
 
-    # We're definitely going to load the templates. We set templates_loaded
-    # here instead of at the end of the file to protect against cyclical
-    # dependencies that could kill the recursive lookup below.
-    this_module.__templates_loaded = True
-
```

Comment:
@thewtex @brad-t-moore I would like to get your feedback on impact from this change. We were previously setting `__templates_loaded = True` when a module was beginning to be loaded, but before we had verified and loaded its module dependencies. The stated intent here for setting this flag early was to protect against cyclical dependencies. In a sense it did this successfully, but as a result I observed errors in loading modules defining factories.

For background, this PR intends to update module loading so that a module defining a base class that is overridden with factories in other modules will load those other modules immediately after it itself is loaded. For instance, the `ITKIOImageBase` defines the base class `ImageIOBase` while the `ITKIOBMP` module defines the overriding class `BMPImageIO`. When `ITKIOImageBase` loads we expect modules to load in the following order:
- Dependencies: ITKPyBase, ITKCommon. Circular dependencies are explicitly disallowed.
- The module itself: load `ITKIOImageBase` template symbols into the Python namespace.
- Factory definitions: Any modules defining ImageIO overrides, such as `ITKIOBMP`.

This works for most cases without changes to line 86 here. However, when we try to directly load a class from `ITKIOBMP` without doing anything else, i.e.
```py
>>> import itk
>>> import itk.BMPImageIO
``` 
we get an error. We can trace the issue back to how we are loading modules and `__templates_loaded`:
1. we start loading `ITKIOBMP`. `__templates_loaded` is defined as `True`. We see that `ITKIOBMP` depends on `ITKPyBase` and `ITKIOImageBase`. We start loading each dependency sequentially.
2. `ITKPyBase` has no dependencies and requires no factory overrides; load succeeds.
3. `ITKIOImageBase` depends on `ITKPyBase` and `ITKCommon`. Both are successfully loaded.
4. `ITKIOImageBase` is successfully loaded.
5. `ITKIOImageBase` requires ImageIO factories. We see that `ITKIOBMP` has `__templates_loaded`, so its templates must be available. We try to get `BMPImageIOFactory`, but the symbol has not _actually_ been loaded yet because at the top level we think we are still loading dependencies. An error occurs and the load fails out.

This issue is avoided by moving the definition of `__templates_loaded` to after templates are actually loaded (which also makes things easier to read). With this change applied to the above scenario, `ITKIOBMP` is actually loaded when `ITKIOImageBase` loads its factories, so when we get back to the top level after loading dependencies we can exit early knowing that `ITKIOBMP` template symbols are already available.

What I'm still unclear on, is the scenario that prompted `__templates_loaded` to be placed before the template loading procedure. Is this still applicable / what cyclical dependency issues are we attempting to guard against? If this is an inadequate solution we could explore two flags, `__templates_initialized` before and `__templates_loaded` after the template loading procedure.

Reply:
I don't think this should be changed -- there is going to be a circular import with derived -> base -> derived. Loading factories during `ITKImageIOBase` loading will cause circular import issues.

Reply:
Thanks @thewtex . I've updated discussion further down to reflect updates for accommodating circular imports with eager loading as well as lazy, which do require moving `__templates_loaded` and adding an extra check to optionally exit early if a circular import results in the module being loaded early. Closing this in favor of that discussion, please feel free to reopen a comment here if you would like to revisit.